doc/Running and Writing Tests for Sia.md

```diff
@@ -0,0 +1,278 @@
+`_test.go` files) for each package, setting off a panic for any test that runs
+longer than 5s.  For verbose output, run `make test-v` (which panics after 15s
+instead of 5s).  Finally, `make test-long` has verbose output, only panics when
+a test takes 5 minutes, and also cleans up your code using `gofmt` and `golint`.  
+**You should run** `make test-long` **before each pull request.**
+
+Run `make cover` to run all tests for each package and generate color-coded
+.html visualizations of test coverage by function for each source file.  Open
+`cover/<module>.html` in a browser to inspect a module's test coverage.
+
+Meanwhile, `make bench` will call `gofmt` on all packages, then run all
+benchmarks (functions starting with `Benchmark` in `_test.go` files).
+
+<a name="particular"/>
+### Testing a particular package
+It can take a while to run the custom `make` commands defined in Sia's Makefile,
+since they check all packages.  If you just want to check the package you've
+made changes in, you can run the commands below.  But don't forget to run `make
+test-long` before making a pull request!
+
+```bash
+# Make sure you are in the right directory.
+$ cd $GOPATH/src/github.com/<your Github username>/Sia
+# Run all tests in directory <package> and view test coverage for <package>.
+$ go test -tags='testing debug' -timeout=360s -covermode=atomic -coverprofile=cover/<package>.out ./<package>
+$ go tool cover -html=cover/<package>.out -o=cover/<package>.html
+$ rm cover/<package>.out
+```
+Now you can open `cover/<package>.html` in a browser to view test coverage for
+<package>.  On OS X, you can just do `open cover/<package.html>`.
+
+<a name="write"/>
+## Writing new tests for Sia
+When you run `make cover`, you'll notice that many files have pretty low
+coverage -- or even none at all.  We're working on fixing that, but we could
+use your help.
+
+<a name="naming"/>
+### A few guidelines
+* The test functions for `filename.go` should go in `filename_test.go` in the
+    same directory.
+* A test function name should start with `Test` and clearly convey what is
+    being tested.
+* You should declare function-specific variables and constants locally (inside
+    the test function) instead of globally (outside the test function).  That
+    holds for all functions.
+* As always, code should adhere to the standards and conventions laid out in
+    [doc/Developers.md][developers]
+
+<a name="basic"/>
+### Basic test format
+Suppose we want to test a function `Foo` in `bar.go` in package `foobar` that
+takes a signed 32-bit integer as argument and returns a result and an error.
+(As you've probably noticed, Go functions can return multiple values.)  
+Let's make a test file named `bar_test.go` in `foobar`.
+
+```go
+//
+package foobar
+
+import (
+    "testing"
+)
+
+func TestFoo(t *testing.T) {
+	// Define a slice of test inputs.
+	testInputs := []int32{0, 1, -1, 42, 1234, -603}
+	// Loop over testInputs.
+	for _, in := range testInputs {
```

Comment:
this is basically a table-driven test. Perhaps for better contrast with the following section, this example should be a different sort of test. Maybe something like this:

```go
func TestFoo(t *testing.T) {
    foo, err := NewFoo()
    if err != nil {
        // if NewFoo failed, we can't continue testing
        t.Fatal(err)
    }

    // try a normal input; should succeed
    err := foo.Bar(3)
    if err != nil {
        // report the error, but don't abort the test
        t.Error(err)
    }

    // try a bad input; should return an error
    // NOTE: always prefer to compare to a specific error, rather than err == nil
    err = foo.Bar(0)
    if err != errDivideByZero {
        t.Errorf("expected errDivideByZero, got %v", errDivideByZero, err)
    }
}
```